File: backend/src/models/PaymentPlan.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IPaymentScheduleItem {
  dueDate: Date;
  amount: number;
  paidDate?: Date;
  status: 'Unpaid' | 'Paid' | 'Overdue';
}

export interface IPaymentPlan extends Document {
  customerId: mongoose.Types.ObjectId;
  totalAmount: number;
  emiAmount: number;
  frequency: 'Monthly' | 'Weekly';
  installmentsCount: number;
  paidInstallmentsCount: number;
  startDate: Date;
  status: 'Active' | 'Completed' | 'Defaulted';
  schedule: IPaymentScheduleItem[];
  createdAt: Date;
  updatedAt: Date;
}

const PaymentPlanSchema: Schema = new Schema(
  {
    customerId: { type: Schema.Types.ObjectId, ref: 'Customer', required: true, index: true },
    totalAmount: { type: Number, required: true, min: 0 },
    emiAmount: { type: Number, required: true, min: 0 },
    frequency: { type: String, enum: ['Monthly', 'Weekly'], default: 'Monthly' },
    installmentsCount: { type: Number, required: true, min: 1 },
    paidInstallmentsCount: { type: Number, default: 0, min: 0 },
    startDate: { type: Date, required: true },
    status: { type: String, enum: ['Active', 'Completed', 'Defaulted'], default: 'Active', index: true },
    schedule: [
      {
        dueDate: { type: Date, required: true },
        amount: { type: Number, required: true },
        paidDate: { type: Date },
        status: { type: String, enum: ['Unpaid', 'Paid', 'Overdue'], default: 'Unpaid' },
      },
    ],
  },
  { timestamps: true }
);

export default mongoose.model<IPaymentPlan>('PaymentPlan', PaymentPlanSchema);

File: backend/models/payment_plan.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class PaymentScheduleItem(BaseModel):
    due_date: datetime
    amount: float
    paid_date: Optional[datetime] = None
    status: str = "Unpaid"  # Unpaid | Paid | Overdue


class PaymentPlanCreate(BaseModel):
    customer_id: str
    total_amount: float
    emi_amount: float
    frequency: str = "Monthly"
    installments_count: int
    start_date: datetime
    schedule: List[PaymentScheduleItem] = []


class PaymentPlanResponse(PaymentPlanCreate):
    id: str
    paid_installments_count: int = 0
    status: str = "Active"  # Active | Completed | Defaulted
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
